examples/syscall-ftrace.c

```diff
@@ -0,0 +1,222 @@
+/**
+ * syscall-ftrace.c
+ * 
+ * System call "stealing" with ftrace
+ * 
+ * We create a callback function that contains
+ * an unconditional jump to our spying function,
+ * which will then return control to the original one.
+ * 
+ * The callback function is triggered by ftrace.
+ */
+
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/moduleparam.h>
+#include <linux/version.h>
+#include <linux/unistd.h>
+#include <linux/kprobes.h>
+#include <linux/sched.h>
+#include <linux/uaccess.h>
+#include <linux/slab.h>
+/** This is what we're using here. */
+#include <linux/ftrace.h>
+
+MODULE_LICENSE("GPL");
+
+#define MAX_FILENAME_SIZE 200
+
+/* UID we want to spy on - will be filled from the command line. */
+static int uid = 0;
+module_param(uid, int, 0644);
+
+/**
+ * This is a housekeeping structure that saves all information
+ * needed for hooking. Usage with `PREPARE_HOOK` is recommended.
+ * 
+ * Example:
+ * static ftrace_hook_t sys_clone_hook = 
+ *      PREPARE_HOOK(__NR_clone, my_sys_clone, &orig_sys_clone)
+ */
+typedef struct ftrace_hook {
+    unsigned long nr; // syscall name
+    void *new; // hook function
+    void *orig; // original function
+
+    unsigned long address; // address to the original function
+    struct ftrace_ops ops; // ftrace structure
+} ftrace_hook_t;
+
+// clang-format off
+#define PREPARE_HOOK(_nr, _hook, _orig)              \
+    {                                                \
+        .nr = (_nr), .new = (_hook), .orig = (_orig) \
+    }
+
+unsigned long **sys_call_table;
+// clang-format on
+
+/**
+ * For the sake of simplicity, only the kprobe method is included.
+ * If you want to know more about different methods to get
+ * kallsyms_lookup_name, see syscall.c.
+ */
+static int resolve_address(ftrace_hook_t *hook)
+{
+    static struct kprobe kp = { .symbol_name = "kallsyms_lookup_name" };
+    unsigned long (*kallsyms_lookup_name)(const char *name);
+    register_kprobe(&kp);
+    kallsyms_lookup_name = (unsigned long (*)(const char *))kp.addr;
+    unregister_kprobe(&kp);
+
+    if (kallsyms_lookup_name)
+        pr_info("[syscall-ftrace] kallsyms_lookup_name is found at 0x%lx\n",
```

Comment:
For `[syscall-ftrace]` prefix, you can use the following macro to make the code cleaner.
```cpp
#undef pr_fmt
#define pr_fmt(fmt) "[syscall-ftrace] " fmt
```